server/src/main/java/org/opensearch/index/document/SortedUnsignedLongDocValuesRangeQuery.java

```diff
@@ -0,0 +1,171 @@
+        this.lowerValue = lowerValue.longValue();
+        this.upperValue = upperValue.longValue();
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (sameClassAs(obj) == false) {
+            return false;
+        }
+        SortedUnsignedLongDocValuesRangeQuery that = (SortedUnsignedLongDocValuesRangeQuery) obj;
+        return Objects.equals(field, that.field) && lowerValue == that.lowerValue && upperValue == that.upperValue;
+    }
+
+    @Override
+    public int hashCode() {
+        int h = classHash();
+        h = 31 * h + field.hashCode();
+        h = 31 * h + Long.hashCode(lowerValue);
+        h = 31 * h + Long.hashCode(upperValue);
+        return h;
+    }
+
+    @Override
+    public void visit(QueryVisitor visitor) {
+        if (visitor.acceptField(field)) {
+            visitor.visitLeaf(this);
+        }
+    }
+
+    @Override
+    public String toString(String field) {
+        StringBuilder b = new StringBuilder();
+        if (this.field.equals(field) == false) {
+            b.append(this.field).append(":");
+        }
+        return b.append("[")
+            .append(Long.toUnsignedString(lowerValue))
+            .append(" TO ")
+            .append(Long.toUnsignedString(upperValue))
+            .append("]")
+            .toString();
+    }
+
+    @Override
+    public Query rewrite(IndexReader reader) throws IOException {
+        if (Long.compareUnsigned(lowerValue, Numbers.MIN_UNSIGNED_LONG_VALUE_AS_LONG) == 0
+            && Long.compareUnsigned(upperValue, Numbers.MAX_UNSIGNED_LONG_VALUE_AS_LONG) == 0) {
+            return new FieldExistsQuery(field);
+        }
+        return super.rewrite(reader);
+    }
+
+    abstract SortedNumericDocValues getValues(LeafReader reader, String field) throws IOException;
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, ScoreMode scoreMode, float boost) throws IOException {
+        return new ConstantScoreWeight(this, boost) {
+
+            @Override
+            public boolean isCacheable(LeafReaderContext ctx) {
+                return DocValues.isCacheable(ctx, field);
+            }
+
+            @Override
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+                SortedNumericDocValues values = getValues(context.reader(), field);
+                if (values == null) {
+                    return null;
+                }
+                final NumericDocValues singleton = DocValues.unwrapSingleton(values);
+                final TwoPhaseIterator iterator;
+                if (singleton != null) {
+                    iterator = new TwoPhaseIterator(singleton) {
+                        @Override
+                        public boolean matches() throws IOException {
+                            final long value = singleton.longValue();
+                            return Long.compareUnsigned(value, lowerValue) >= 0 && Long.compareUnsigned(value, upperValue) <= 0;
```

Comment:
I just saw the code, search_after is not supported for unsigned_long as of now.

Reply:
Hm ... it should be supported, I do have a test case for it [1]

> I just saw the code, search_after is not supported for unsigned_long as of now.

Could you please point me the place that I missed, the `search_after` should be supported I think, thank you.

[1] https://github.com/opensearch-project/OpenSearch/pull/6237/files#diff-dd44653f1a2055a1e3cdedaac251645e4a9d2cf3170cd017e8fe96723cc416cd

Reply:
Somehow I overlooked here we were throwing exception in searchafter builder in case of unsigned_log
server/src/main/java/org/opensearch/search/searchafter/SearchAfterBuilder.java

All good ! Thank you for clarifying.